util/coordinator.py

```diff
@@ -0,0 +1,706 @@
+                        # Second try to get a next job
+                        job = self._get_job(worker)
+                        if job is None:
+                            # Albeit the epoch state machine got a new epoch, the epoch had no new job for us
+                            log_error('Unexpected case - no job for worker %d.' % (worker))
+                        return job
+
+                    # Epoch state machine has no new epoch
+                    # This happens at the end of the whole training - nothing to worry about
+                    log_traffic('No jobs left for worker %d.' % (worker))
+                    self._log_all_jobs()
+                    return None
+
+                # We got a new job from one of the currently running epochs
+                log_traffic('Got new %s' % job)
+                return job
+
+            # We are a remote worker and have to hand over to the chief worker by HTTP
+            result = self._talk_to_chief(PREFIX_GET_JOB + str(FLAGS.task_index))
+            if result:
+                result = pickle.loads(result)
+            return result
+
+    def next_job(self, job):
+        '''Sends a finished job back to the coordinator and retrieves in exchange the next one.
+
+        Kwargs:
+            job (WorkerJob): job that was finished by a worker and who's results are to be
+                digested by the coordinator
+
+        Returns:
+            WorkerJob. next job of one of the running epochs that will get
+                associated with the worker from the finished job and put into state 'running'
+        '''
+        if self.is_chief:
+            # Try to find the epoch the job belongs to
+            epoch = next((epoch for epoch in self._epochs_running if epoch.id == job.epoch_id), None)
+            if epoch:
+                # We are going to manipulate things - let's avoid undefined state
+                with self._lock:
+                    # Let the epoch finish the job
+                    epoch.finish_job(job)
+                    # Check, if epoch is done now
+                    if epoch.done():
+                        # If it declares itself done, move it from 'running' to 'done' collection
+                        self._epochs_running.remove(epoch)
+                        self._epochs_done.append(epoch)
+                        log_info('%s' % epoch)
+            else:
+                # There was no running epoch found for this job - this should never happen.
+                log_error('There is no running epoch of ID %d for job with ID %d. This should never happen.' % (job.epoch_id, job.id))
+            return self.get_job(job.worker)
+
+        # We are a remote worker and have to hand over to the chief worker by HTTP
+        result = self._talk_to_chief('', data=pickle.dumps(job))
+        if result:
+            result = pickle.loads(result)
+        return result
+
+class GlobalConfig:
```

Comment:
This should not live here in the coordinator. You could also somehow join it with the flags module.

Reply:
There's a circular dependency between the globals and the coordinator (the COORD global) and IMO breaking it by separating just that global initialization out of `initialize_globals` would be worse than having these live in the same place. What do you think?

Reply:
I'd initalize the `COORD` global variable in `DeepSpeech.py` (moving the block with comment `# Initializing and starting the training coordinator` back there). Then there should be a clean separation of concerns again.